File: apps/ios/Cursumi/Core/Models/Quiz.swift
import Foundation

// Quizzes, examen y minijuegos. Misma lógica de calificación que el servidor.

enum QuizQuestionType: String, Decodable {
    case multipleChoice = "multiple-choice"
    case trueFalse = "true-false"
    case checkbox
    case ordering
    case matching
    case shortAnswer = "short-answer"
}

/// Respuesta del alumno, con la misma forma que espera el servidor:
///  - `.index`   → opción múltiple / verdadero-falso (índice)
///  - `.indices` → casillas (índices marcados)
///  - `.texts`   → ordenar (textos en el orden elegido) / relacionar (pareja
///                 elegida para cada elemento izquierdo, en el orden de `options`)
enum QuizAnswer: Encodable, Equatable {
    case index(Int)
    case indices([Int])
    case texts([String])

    func encode(to encoder: Encoder) throws {
        var c = encoder.singleValueContainer()
        switch self {
        case .index(let i): try c.encode(i)
        case .indices(let a): try c.encode(a)
        case .texts(let a): try c.encode(a)
        }
    }

    /// Una respuesta cuenta si no está vacía ni tiene huecos.
    var isAnswered: Bool {
        switch self {
        case .index: return true
        case .indices(let a): return !a.isEmpty
        case .texts(let a): return !a.isEmpty && a.allSatisfy { !$0.isEmpty }
        }
    }
}

struct QuizQuestion: Equatable {
    var question: String
    /// ordenar: opciones en su orden CORRECTO. relacionar: columna izquierda.
    var options: [String]
    var type: QuizQuestionType = .multipleChoice
    var correctAnswer: Int?
    var correctAnswers: [Int]?
    /// relacionar: `matchRight[i]` es la pareja correcta de `options[i]`.
    var matchRight: [String]?

    /// Califica una respuesta con los mismos criterios que el servidor.
    func grade(_ answer: QuizAnswer?) -> Bool {
        guard let answer else { return false }
        switch type {
        case .checkbox:
            guard case .indices(let sel) = answer, let expected = correctAnswers else { return false }
            return Set(sel) == Set(expected)
        case .ordering:
            guard case .texts(let a) = answer else { return false }
            return a == options
        case .matching:
            guard case .texts(let a) = answer, let right = matchRight else { return false }
            return a == right
        default:
            guard case .index(let i) = answer else { return false }
            return i == correctAnswer
        }
    }

    private init(question: String, options: [String], type: QuizQuestionType, correctAnswer: Int?, correctAnswers: [Int]?, matchRight: [String]?) {
        self.question = question
        self.options = options
        self.type = type
        self.correctAnswer = correctAnswer
        self.correctAnswers = correctAnswers
        self.matchRight = matchRight
    }

    static func make(question: String, options: [String], type: QuizQuestionType = .multipleChoice, correctAnswer: Int? = nil, correctAnswers: [Int]? = nil, matchRight: [String]? = nil) -> QuizQuestion {
        QuizQuestion(question: question, options: options, type: type, correctAnswer: correctAnswer, correctAnswers: correctAnswers, matchRight: matchRight)
    }

    /// Preguntas de un quiz de lección (`lesson.content` es JSON: array o `{questions}`).
    static func parseLessonQuiz(_ content: String?) -> [QuizQuestion] {
        guard let content, let json = JSONValue.parse(content) else { return [] }
        return parse(list: json.array ?? json["questions"]?.array, correctKey: "correctAnswer")
    }

    /// Preguntas del quiz de sección (`section.quiz`, Json de Prisma).
    static func parseSectionQuiz(_ raw: JSONValue?) -> [QuizQuestion] {
        guard let raw else { return [] }
        return parse(list: raw.array ?? raw["questions"]?.array, correctKey: "correct")
            .filter { !$0.options.isEmpty }
    }

    private static func parse(list: [JSONValue]?, correctKey: String) -> [QuizQuestion] {
        (list ?? []).map { q in
            QuizQuestion(
                question: q["question"]?.string ?? "",
                options: q["options"]?.stringArray ?? [],
                type: q["type"]?.string.flatMap(QuizQuestionType.init) ?? .multipleChoice,
                correctAnswer: q[correctKey]?.int ?? (correctKey == "correct" ? 0 : nil),
                correctAnswers: q["correctAnswers"]?.intArray,
                matchRight: q["matchRight"]?.stringArray
            )
        }
    }
}

struct QuizConfig: Equatable {
    /// Límite en minutos (0 = sin límite).
    var timeLimitMin = 0
    /// Máximo de intentos (0 = ilimitados).
    var maxAttempts = 0
    /// Puntaje mínimo para aprobar (0–100).
    var passingScore = 70

    static func parse(_ content: String?) -> QuizConfig {
        guard let content, let json = JSONValue.parse(content), case .object = json else { return QuizConfig() }
        return QuizConfig(
            timeLimitMin: json["timeLimit"]?.int ?? 0,
            maxAttempts: json["attempts"]?.int ?? 0,
            passingScore: json["passingScore"]?.int ?? 70
        )
    }
}

// MARK: - Examen final

struct ExamQuestion: Decodable, Identifiable {
    let id: String
    let question: String
    let type: QuizQuestionType?
    /// ordenar: ya barajadas por el servidor. relacionar: columna izquierda.
    let options: [String]?
    /// relacionar: columna derecha barajada, sin revelar la pareja.
    let matchRight: [String]?
    let points: Double?

    var isGradable: Bool { !(options ?? []).isEmpty }
}

struct Exam: Decodable {
    let id: String
    let passingScore: Int
    let timeLimit: Int?
    let questions: [ExamQuestion]
}

struct ExamResult: Decodable {
    struct Certificate: Decodable { let id: String; let number: String }
    let score: Double
    let passed: Bool
    let certificate: Certificate?
}

// MARK: - Minijuegos de sección

enum Minigame: Equatable {
    struct Pair: Equatable { let term: String; let definition: String }
    struct Word: Equatable { let word: String; let hint: String }
    struct MatchPair: Equatable { let left: String; let right: String }

    case memory(instruction: String?, pairs: [Pair])
    case hangman(instruction: String?, words: [Word])
    case sort(instruction: String?, items: [String])
    case match(instruction: String?, pairs: [MatchPair])

    var instruction: String? {
        switch self {
        case .memory(let i, _), .hangman(let i, _), .sort(let i, _), .match(let i, _): return i
        }
    }

    static func parse(_ raw: JSONValue?) -> Minigame? {
        guard let raw, let type = raw["type"]?.string else { return nil }
        let instruction = raw["instruction"]?.string
        switch type {
        case "memory":
            let pairs = (raw["pairs"]?.array ?? []).compactMap { p -> Pair? in
                guard let t = p["term"]?.string, let d = p["definition"]?.string else { return nil }
                return Pair(term: t, definition: d)
            }
            return pairs.isEmpty ? nil : .memory(instruction: instruction, pairs: pairs)
        case "hangman":
            let words = (raw["words"]?.array ?? []).compactMap { w -> Word? in
                guard let word = w["word"]?.string else { return nil }
                return Word(word: word, hint: w["hint"]?.string ?? "")
            }
            return words.isEmpty ? nil : .hangman(instruction: instruction, words: words)
        case "sort":
            let items = raw["items"]?.stringArray ?? []
            return items.isEmpty ? nil : .sort(instruction: instruction, items: items)
        case "match":
            let pairs = (raw["pairs"]?.array ?? []).compactMap { p -> MatchPair? in
                guard let l = p["left"]?.string, let r = p["right"]?.string else { return nil }
                return MatchPair(left: l, right: r)
            }
            return pairs.isEmpty ? nil : .match(instruction: instruction, pairs: pairs)
        default:
            return nil
        }
    }
}

/// Quita acentos y pasa a mayúsculas para comparar letras en el ahorcado. La Ñ se
/// conserva como letra propia (el teclado del juego la tiene).
func normalizedLetters(_ s: String) -> String {
    s.map { ch -> String in
        if ch == "ñ" || ch == "Ñ" { return "Ñ" }
        return String(ch).folding(options: .diacriticInsensitive, locale: nil).uppercased()
    }.joined()
}

extension Int {
    /// Segundos → "m:ss".
    var mmss: String { String(format: "%d:%02d", self / 60, self % 60) }
}

File: apps/android/app/src/main/java/com/cursumi/app/core/model/Quiz.kt
package com.cursumi.app.core.model

import com.cursumi.app.core.api.AppJson
import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonEncoder
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonPrimitive

// Quizzes, examen y minijuegos. Espejo de la lógica de `apps/mobile/src/lib/me.ts`.

enum class QuizQuestionType(val wire: String) {
    MULTIPLE_CHOICE("multiple-choice"), TRUE_FALSE("true-false"), CHECKBOX("checkbox"),
    ORDERING("ordering"), MATCHING("matching"), SHORT_ANSWER("short-answer");

    companion object {
        fun from(raw: String?): QuizQuestionType = entries.firstOrNull { it.wire == raw } ?: MULTIPLE_CHOICE
    }
}

/**
 * Respuesta del alumno, con la forma que espera el servidor:
 *  - [Index]   → opción múltiple / verdadero-falso (índice)
 *  - [Indices] → casillas (índices marcados)
 *  - [Texts]   → ordenar (textos en el orden elegido) / relacionar (pareja
 *                elegida para cada elemento izquierdo, en el orden de `options`)
 */
@Serializable(with = QuizAnswerSerializer::class)
sealed interface QuizAnswer {
    data class Index(val value: Int) : QuizAnswer
    data class Indices(val values: List<Int>) : QuizAnswer
    data class Texts(val values: List<String>) : QuizAnswer

    /** Una respuesta cuenta si no está vacía ni tiene huecos. */
    val isAnswered: Boolean
        get() = when (this) {
            is Index -> true
            is Indices -> values.isNotEmpty()
            is Texts -> values.isNotEmpty() && values.all { it.isNotEmpty() }
        }
}

object QuizAnswerSerializer : KSerializer<QuizAnswer> {
    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("QuizAnswer")
    override fun serialize(encoder: Encoder, value: QuizAnswer) {
        val el: JsonElement = when (value) {
            is QuizAnswer.Index -> JsonPrimitive(value.value)
            is QuizAnswer.Indices -> buildJsonArray { value.values.forEach { add(JsonPrimitive(it)) } }
            is QuizAnswer.Texts -> buildJsonArray { value.values.forEach { add(JsonPrimitive(it)) } }
        }
        (encoder as JsonEncoder).encodeJsonElement(el)
    }
    override fun deserialize(decoder: Decoder): QuizAnswer = throw UnsupportedOperationException()
}

data class QuizQuestion(
    val question: String,
    /** ordenar: opciones en su orden CORRECTO. relacionar: columna izquierda. */
    val options: List<String>,
    val type: QuizQuestionType = QuizQuestionType.MULTIPLE_CHOICE,
    val correctAnswer: Int? = null,
    val correctAnswers: List<Int>? = null,
    /** relacionar: `matchRight[i]` es la pareja correcta de `options[i]`. */
    val matchRight: List<String>? = null,
) {
    /** Califica con los mismos criterios que el servidor. */
    fun grade(answer: QuizAnswer?): Boolean {
        if (answer == null) return false
        return when (type) {
            QuizQuestionType.CHECKBOX -> answer is QuizAnswer.Indices && correctAnswers != null && answer.values.toSet() == correctAnswers.toSet()
            QuizQuestionType.ORDERING -> answer is QuizAnswer.Texts && answer.values == options
            QuizQuestionType.MATCHING -> answer is QuizAnswer.Texts && matchRight != null && answer.values == matchRight
            else -> answer is QuizAnswer.Index && answer.value == correctAnswer
        }
    }

    companion object {
        /** Preguntas de un quiz de lección (`lesson.content` es JSON: array o `{questions}`). */
        fun parseLessonQuiz(content: String?): List<QuizQuestion> {
            val json = parseJson(content) ?: return emptyList()
            return parse(json.listOrQuestions(), correctKey = "correctAnswer")
        }

        /** Preguntas del quiz de sección (`section.quiz`, Json de Prisma). */
        fun parseSectionQuiz(raw: JsonElement?): List<QuizQuestion> =
            parse(raw?.listOrQuestions(), correctKey = "correct").filter { it.options.isNotEmpty() }

        private fun JsonElement.listOrQuestions(): JsonArray? =
            this as? JsonArray ?: (this as? JsonObject)?.get("questions") as? JsonArray

        private fun parse(list: JsonArray?, correctKey: String): List<QuizQuestion> = (list ?: emptyList()).mapNotNull { q ->
            val o = q as? JsonObject ?: return@mapNotNull null
            QuizQuestion(
                question = o.str("question") ?: "",
                options = o.strList("options"),
                type = QuizQuestionType.from(o.str("type")),
                correctAnswer = o.int(correctKey) ?: if (correctKey == "correct") 0 else null,
                correctAnswers = (o["correctAnswers"] as? JsonArray)?.mapNotNull { (it as? JsonPrimitive)?.intOrNull },
                matchRight = (o["matchRight"] as? JsonArray)?.let { a -> a.mapNotNull { (it as? JsonPrimitive)?.content } },
            )
        }
    }
}

fun parseJson(text: String?): JsonElement? = text?.let { runCatching { AppJson.parseToJsonElement(it) }.getOrNull() }
fun JsonObject.str(key: String): String? = (this[key] as? JsonPrimitive)?.takeIf { it.isString || it.content != "null" }?.content
fun JsonObject.int(key: String): Int? = (this[key] as? JsonPrimitive)?.let { it.intOrNull ?: it.content.toDoubleOrNull()?.toInt() }
fun JsonObject.strList(key: String): List<String> = (this[key] as? JsonArray)?.mapNotNull { (it as? JsonPrimitive)?.content } ?: emptyList()

data class QuizConfig(val timeLimitMin: Int = 0, val maxAttempts: Int = 0, val passingScore: Int = 70) {
    companion object {
        fun parse(content: String?): QuizConfig {
            val o = parseJson(content) as? JsonObject ?: return QuizConfig()
            return QuizConfig(o.int("timeLimit") ?: 0, o.int("attempts") ?: 0, o.int("passingScore") ?: 70)
        }
    }
}

// ── Examen final ─────────────────────────────────────────────────────────────

@Serializable
data class ExamQuestion(
    val id: String,
    val question: String,
    val type: String? = null,
    /** ordenar: ya barajadas por el servidor. relacionar: columna izquierda. */
    val options: List<String>? = null,
    /** relacionar: columna derecha barajada. */
    val matchRight: List<String>? = null,
    val points: Double? = null,
) {
    val isGradable get() = !options.isNullOrEmpty()
    val questionType get() = QuizQuestionType.from(type)
}

@Serializable data class Exam(val id: String, val passingScore: Int = 70, val timeLimit: Int? = null, val questions: List<ExamQuestion> = emptyList())
@Serializable data class ExamResult(val score: Double = 0.0, val passed: Boolean = false, val certificate: Cert? = null) {
    @Serializable data class Cert(val id: String, val number: String)
}

// ── Minijuegos ───────────────────────────────────────────────────────────────

sealed interface Minigame {
    val instruction: String?
    data class Pair(val term: String, val definition: String)
    data class Word(val word: String, val hint: String)
    data class MatchPair(val left: String, val right: String)

    data class Memory(override val instruction: String?, val pairs: List<Pair>) : Minigame
    data class Hangman(override val instruction: String?, val words: List<Word>) : Minigame
    data class Sort(override val instruction: String?, val items: List<String>) : Minigame
    data class Match(override val instruction: String?, val pairs: List<MatchPair>) : Minigame

    companion object {
        fun parse(raw: JsonElement?): Minigame? {
            val o = raw as? JsonObject ?: return null
            val instruction = o.str("instruction")
            return when (o.str("type")) {
                "memory" -> (o["pairs"] as? JsonArray)?.mapNotNull { p -> (p as? JsonObject)?.let { po -> po.str("term")?.let { t -> po.str("definition")?.let { d -> Pair(t, d) } } } }
                    ?.takeIf { it.isNotEmpty() }?.let { Memory(instruction, it) }
                "hangman" -> (o["words"] as? JsonArray)?.mapNotNull { w -> (w as? JsonObject)?.let { wo -> wo.str("word")?.let { Word(it, wo.str("hint") ?: "") } } }
                    ?.takeIf { it.isNotEmpty() }?.let { Hangman(instruction, it) }
                "sort" -> o.strList("items").takeIf { it.isNotEmpty() }?.let { Sort(instruction, it) }
                "match" -> (o["pairs"] as? JsonArray)?.mapNotNull { p -> (p as? JsonObject)?.let { po -> po.str("left")?.let { l -> po.str("right")?.let { r -> MatchPair(l, r) } } } }
                    ?.takeIf { it.isNotEmpty() }?.let { Match(instruction, it) }
                else -> null
            }
        }
    }
}
